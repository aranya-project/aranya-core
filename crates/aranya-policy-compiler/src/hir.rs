//! A High-level Intermediate Representation (HIR) for Aranya
//! policies.

#![allow(dead_code)]
#![allow(clippy::module_inception)]

mod arena;
mod dsl;
mod hir;
mod lower;
mod snapshot_tests;
mod visit;

use aranya_policy_ast::Policy;
use aranya_policy_module::ffi::ModuleSchema;

use crate::hir::lower::LowerCtx;
pub(crate) use crate::hir::{arena::AstNodes, hir::*, visit::Visitor};

/// Parses a [`Policy`] into [`Hir`] and [`AstNodes`].
pub(crate) fn parse<'a>(
    policy: &'a Policy,
    ffi_modules: &'a [ModuleSchema<'a>],
) -> (Hir, AstNodes<'a>) {
    let LowerCtx { ast, arena } = LowerCtx::build(policy, ffi_modules);
    (ast, arena)
}
