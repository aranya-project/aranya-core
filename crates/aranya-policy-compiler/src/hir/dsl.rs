//! Provides a convenient macro for constructing [`Hir`] in
//! tests.

#![cfg(test)]

use slotmap::KeyData;

/// Creates an ID from an index and version using slotmap's KeyData.
///
/// # Panics
/// - If idx is 0
/// - If version is 0
pub(crate) fn make_id<T: From<KeyData>>(idx: u32, version: u32) -> T {
    assert!(idx > 0, "idx must be greater than 0");
    assert!(version > 0, "version must be greater than 0");
    let v = (idx as u64) | ((version as u64) << 32);
    KeyData::from_ffi(v).into()
}
