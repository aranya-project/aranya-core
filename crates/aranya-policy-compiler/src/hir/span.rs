use core::{
    cmp::Ordering,
    fmt,
    ops::{Bound, Range, RangeBounds},
};

use serde::{Deserialize, Serialize};

/// A range in the source text.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Span {
    // [start, end)
    start: usize,
    end: usize,
}

impl Span {
    /// Create a new span
    pub fn new(start: usize, end: usize) -> Self {
        debug_assert!(
            start <= end,
            "invalid span: start ({}) must be <= end ({})",
            start,
            end
        );
        Span { start, end }
    }

    /// Returns the start position.
    pub fn start(&self) -> usize {
        self.start
    }

    /// Returns the end position.
    pub fn end(&self) -> usize {
        self.end
    }

    /// Reports whether `self` contains `other`.
    pub fn contains<U>(&self, other: U) -> bool
    where
        U: RangeBounds<usize>,
    {
        self.intersect(&other.to_bounds()) == other.to_bounds() && !other.is_empty()
    }

    /// Merges two spans into a single span.
    pub fn merge(&self, other: Span) -> Span {
        Self::new(self.start.min(other.start), self.end.max(other.end))
    }

    /// Returns the length of the span.
    pub fn len(&self) -> usize {
        self.end.saturating_sub(self.start)
    }

    /// Reports whether the span is empty.
    pub fn is_empty(&self) -> bool {
        self.start >= self.end
    }

    /// Converts the span into a [`Range`].
    pub fn into_range(self) -> Range<usize> {
        self.start..self.end
    }
}

impl Default for Span {
    fn default() -> Self {
        Self::new(0, 0)
    }
}

impl RangeBounds<usize> for Span {
    fn start_bound(&self) -> Bound<&usize> {
        Bound::Included(&self.start)
    }

    fn end_bound(&self) -> Bound<&usize> {
        Bound::Excluded(&self.end)
    }
}

impl RangeBounds<usize> for &Span {
    fn start_bound(&self) -> Bound<&usize> {
        (**self).start_bound()
    }

    fn end_bound(&self) -> Bound<&usize> {
        (**self).end_bound()
    }
}

impl From<Range<usize>> for Span {
    fn from(range: Range<usize>) -> Self {
        Span::new(range.start, range.end)
    }
}

impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.start..span.end
    }
}

impl From<aranya_policy_module::Span<'_>> for Span {
    fn from(span: aranya_policy_module::Span<'_>) -> Self {
        Span::new(span.start(), span.end())
    }
}

impl fmt::Debug for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        Range::from(*self).fmt(f)
    }
}

/// Extension trait for [`RangeBounds`].
///
/// Similar to the +nightly `IntoBounds`.
trait RangeBoundsExt<T>: RangeBounds<T> {
    /// Returns the (start, end) bounds.
    fn to_bounds(&self) -> (Bound<&T>, Bound<&T>) {
        (self.start_bound(), self.end_bound())
    }

    /// Reports whether `self` is the empty range.
    fn is_empty(&self) -> bool
    where
        T: PartialOrd,
    {
        use Bound::*;
        !match (self.start_bound(), self.end_bound()) {
            (Unbounded, _) | (_, Unbounded) => true,
            (Included(start), Excluded(end))
            | (Excluded(start), Included(end))
            | (Excluded(start), Excluded(end)) => start < end,
            (Included(start), Included(end)) => start <= end,
        }
    }

    /// Returns the intersection of `self` and `other`.
    fn intersect<'a, U>(&'a self, other: &'a U) -> (Bound<&'a T>, Bound<&'a T>)
    where
        Self: Sized,
        U: RangeBounds<T>,
        T: Ord,
    {
        use Bound::*;

        let (self_start, self_end) = self.to_bounds();
        let (other_start, other_end) = other.to_bounds();

        let start = match (self_start, other_start) {
            (Included(a), Included(b)) => Included(Ord::max(a, b)),
            (Excluded(a), Excluded(b)) => Excluded(Ord::max(a, b)),
            (Unbounded, Unbounded) => Unbounded,
            (x, Unbounded) | (Unbounded, x) => x,
            (Included(i), Excluded(e)) | (Excluded(e), Included(i)) => {
                if i > e {
                    Included(i)
                } else {
                    Excluded(e)
                }
            }
        };
        let end = match (self_end, other_end) {
            (Included(a), Included(b)) => Included(Ord::min(a, b)),
            (Excluded(a), Excluded(b)) => Excluded(Ord::min(a, b)),
            (Unbounded, Unbounded) => Unbounded,
            (x, Unbounded) | (Unbounded, x) => x,
            (Included(i), Excluded(e)) | (Excluded(e), Included(i)) => {
                if i < e {
                    Included(i)
                } else {
                    Excluded(e)
                }
            }
        };
        (start, end)
    }
}
impl<R, T> RangeBoundsExt<T> for R where R: RangeBounds<T> {}

/// A trait for types that can provide a source span.
pub trait Spanned {
    /// Returns a span covering the contents of the item.
    fn span(&self) -> Span;
}

impl<T: Spanned> Spanned for &T {
    fn span(&self) -> Span {
        (**self).span()
    }
}

impl<A, B> Spanned for (A, B)
where
    A: Spanned,
    B: Spanned,
{
    fn span(&self) -> Span {
        self.0.span().merge(self.1.span())
    }
}

impl<T: Spanned> Spanned for [T] {
    fn span(&self) -> Span {
        self.iter()
            .map(Spanned::span)
            .reduce(|acc, span| acc.merge(span))
            .unwrap_or_default()
    }
}

impl<T: Spanned, const N: usize> Spanned for [T; N] {
    fn span(&self) -> Span {
        self.as_slice().span()
    }
}

impl<T: Spanned> Spanned for Vec<T> {
    fn span(&self) -> Span {
        self.as_slice().span()
    }
}

impl<T: Spanned> Spanned for Option<T> {
    fn span(&self) -> Span {
        self.as_ref().map(Spanned::span).unwrap_or_default()
    }
}

macro_rules! spanned {
    (
        $(#[$meta:meta])*
        $vis:vis struct $name:ident {
            $(
                $(#[$field_meta:meta])*
                $field_vis:vis $field:ident : $ty:ty
            ),+ $(,)?
        }
    ) => {
        $(#[$meta])*
        $vis struct $name {
            $(
                $(#[$field_meta])*
                $field_vis $field: $ty,
            )+
        }
        impl Spanned for $name {
            fn span(&self) -> Span {
                let spans = &[ $(self.$field.span()),* ];
                spans
                    .iter()
                    .copied()
                    .reduce(|acc, span| acc.merge(span))
                    .unwrap_or_default()
            }
        }
    };
}
pub(crate) use spanned;

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! span {
        ($start:expr, $end:expr) => {
            Span::new($start, $end)
        };
    }

    #[test]
    #[should_panic(expected = "invalid span")]
    #[cfg(debug_assertions)]
    fn test_span_new_invalid() {
        // This should panic in debug mode
        span!(10, 5);
    }

    #[test]
    fn test_span_contains() {
        let test_cases = [
            (span!(10, 20), 12..18, true),                 // inner range contained
            (span!(10, 20), 5..25, false),                 // outer range not contained
            (span!(10, 20), 15..25, false),                // overlapping range not contained
            (span!(10, 20), 5..15, false),                 // partial overlap not contained
            (span!(10, 20), 10..20, true),                 // span contains itself
            (span!(10, 20), 10..19, true),                 // starts at boundary, ends before
            (span!(10, 20), 11..20, true),                 // ends at boundary (excluded)
            (span!(10, 20), 10..20, true),                 // exact same range
            (span!(10, 20), 15..15, false),                // empty range inside span
            (span!(10, 20), 10..10, false),                // empty range at start
            (span!(10, 20), 20..20, false),                // empty range at end (excluded)
            (span!(10, 10), 10..10, false), // empty span contains empty range at same position
            (span!(0, usize::MAX), 0..100, true), // max span contains regular range
            (span!(0, usize::MAX), 0..(usize::MAX), true), // max span contains almost-max range
        ];
        for (i, (span, other, want)) in test_cases.iter().enumerate() {
            let got = span.contains(other.clone());
            assert_eq!(got, *want, "#{i}: contains({span:?}, {other:?})");
        }

        let inclusive_cases = [
            (span!(10, 20), 12..=17, true),  // inclusive range contained
            (span!(10, 20), 10..=19, true),  // inclusive range at boundaries
            (span!(10, 20), 10..=20, false), // inclusive range extends to excluded end
            (span!(10, 20), 15..=25, false), // inclusive range extends beyond
            (span!(10, 11), 10..=10, true),  // single-element inclusive range
        ];
        for (i, (span, other, want)) in inclusive_cases.iter().enumerate() {
            let got = span.contains(other.clone());
            assert_eq!(got, *want, "#{i}: contains({span:?}, {other:?})");
        }

        let range_full_cases = [
            (span!(10, 20), false),        // regular span can't contain unbounded range
            (span!(0, usize::MAX), false), // even max span can't contain unbounded range
            (Span::default(), false),      // empty span can't contain unbounded range
        ];
        for (i, (span, want)) in range_full_cases.iter().enumerate() {
            let got = span.contains(..);
            assert_eq!(got, *want, "#{i}: contains({span:?}, ..)");
        }

        let range_from_cases = [
            (span!(10, 20), 15, false),         // unbounded end not contained
            (span!(10, 20), 10, false),         // starts at boundary, unbounded end
            (span!(10, 20), 0, false),          // starts before, unbounded end
            (span!(10, 20), 25, false),         // starts after, unbounded end
            (span!(0, usize::MAX), 100, false), // even max span can't contain unbounded end
            (span!(0, usize::MAX), 0, false),   // max span starting at 0, unbounded end
        ];
        for (i, (span, start, want)) in range_from_cases.iter().enumerate() {
            let got = span.contains(*start..);
            assert_eq!(got, *want, "#{i}: contains({span:?}, {start}..)");
        }

        let range_to_cases = [
            (span!(10, 20), 15, false),                // unbounded start not contained
            (span!(10, 20), 20, false),                // unbounded start, ends at boundary
            (span!(10, 20), 25, false),                // unbounded start, ends after
            (span!(10, 20), 5, false),                 // unbounded start, ends before
            (span!(0, usize::MAX), 100, false), // even max span can't contain unbounded start
            (span!(0, usize::MAX), usize::MAX, false), // max span ending at max, unbounded start
        ];
        for (i, (span, end, want)) in range_to_cases.iter().enumerate() {
            let got = span.contains(..*end);
            assert_eq!(got, *want, "#{i}: contains({span:?}, ..{end})");
        }

        let range_to_inclusive_cases = [
            (span!(10, 20), 15, false),         // unbounded start not contained
            (span!(10, 20), 19, false),         // unbounded start, ends at last valid
            (span!(10, 20), 20, false),         // unbounded start, ends at boundary
            (span!(10, 20), 25, false),         // unbounded start, ends after
            (span!(0, usize::MAX), 100, false), // even max span can't contain unbounded start
            (span!(0, usize::MAX), usize::MAX - 1, false), // max span, unbounded start
        ];
        for (i, (span, end, want)) in range_to_inclusive_cases.iter().enumerate() {
            let got = span.contains(..=*end);
            assert_eq!(got, *want, "#{i}: contains({span:?}, ..={end})");
        }
    }

    #[test]
    fn test_span_merge() {
        let test_cases = [
            (span!(10, 20), span!(30, 40), span!(10, 40)), // non-overlapping spans
            (span!(10, 20), span!(15, 35), span!(10, 35)), // overlapping spans
            (span!(30, 40), span!(10, 20), span!(10, 40)), // merge order doesn't matter
            (span!(10, 20), span!(10, 20), span!(10, 20)), // merging with self
            (span!(0, 10), span!(5, 15), span!(0, 15)),    // partial overlap
            (span!(10, 20), span!(5, 12), span!(5, 20)),   // left extension
            (span!(10, 20), span!(18, 25), span!(10, 25)), // right extension
            (span!(15, 25), span!(10, 30), span!(10, 30)), // fully contained by other
            (span!(10, 30), span!(15, 25), span!(10, 30)), // fully contains other
            (Span::default(), span!(10, 20), span!(0, 20)), // empty with non-empty
            (span!(0, usize::MAX), span!(100, 200), span!(0, usize::MAX)), // max span with regular span
        ];
        for (i, (lhs, rhs, want)) in test_cases.iter().enumerate() {
            let got = lhs.merge(*rhs);
            assert_eq!(got, *want, "#{i}: merge({lhs:?}, {rhs:?})");
        }
    }

    #[test]
    fn test_span_len() {
        let test_cases = [
            (span!(0, 10), 10),                     // regular span length
            (span!(5, 5), 0),                       // empty span length
            (span!(100, 150), 50),                  // larger span length
            (Span::default(), 0),                   // default span length
            (span!(0, 1), 1),                       // single unit span
            (span!(usize::MAX - 1, usize::MAX), 1), // max boundary span
        ];
        for (i, (span, want)) in test_cases.iter().enumerate() {
            let got = span.len();
            assert_eq!(got, *want, "#{i}: len({span:?})");
        }
    }

    #[test]
    fn test_span_is_empty() {
        let test_cases = [
            (span!(0, 10), false),                      // regular span is not empty
            (span!(0, 0), true),                        // zero-length span is empty
            (Span::default(), true),                    // default span is empty
            (span!(5, 5), true),                        // same start/end span is empty
            (span!(1, 2), false),                       // single unit span is not empty
            (span!(usize::MAX - 1, usize::MAX), false), // max boundary span is not empty
        ];
        for (i, (span, want)) in test_cases.iter().enumerate() {
            let got = span.is_empty();
            assert_eq!(got, *want, "#{i}: is_empty({span:?})");
        }
    }
}
