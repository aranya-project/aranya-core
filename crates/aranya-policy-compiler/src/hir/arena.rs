use std::borrow::Cow;

use aranya_policy_ast::{self as ast, AstNode};
use serde::{Deserialize, Serialize};
use slotmap::SecondaryMap;

use crate::hir::hir::{
    ActionArgId, ActionId, BlockId, CmdFieldId, CmdId, EffectFieldId, EffectId, EnumId, ExprId,
    FactId, FactKeyId, FactValId, FinishFuncArgId, FinishFuncId, FuncArgId, FuncId, GlobalId,
    IdentId, StmtId, StructFieldId, StructId, VTypeId,
};

/// An arena of AST nodes.
#[derive(Clone, Default, Debug, Serialize, Deserialize)]
#[cfg_attr(test, derive(PartialEq))]
pub(crate) struct AstNodes<'ast> {
    pub actions: SecondaryMap<ActionId, Cow<'ast, AstNode<ast::ActionDefinition>>>,
    pub action_args: SecondaryMap<ActionArgId, Cow<'ast, ast::FieldDefinition>>,
    pub cmds: SecondaryMap<CmdId, Cow<'ast, AstNode<ast::CommandDefinition>>>,
    pub cmd_fields: SecondaryMap<CmdFieldId, Cow<'ast, ast::StructItem<ast::FieldDefinition>>>,
    pub effects: SecondaryMap<EffectId, Cow<'ast, AstNode<ast::EffectDefinition>>>,
    pub effect_fields:
        SecondaryMap<EffectFieldId, Cow<'ast, ast::StructItem<ast::EffectFieldDefinition>>>,
    pub enums: SecondaryMap<EnumId, Cow<'ast, AstNode<ast::EnumDefinition>>>,
    pub facts: SecondaryMap<FactId, Cow<'ast, AstNode<ast::FactDefinition>>>,
    pub fact_keys: SecondaryMap<FactKeyId, Cow<'ast, ast::FieldDefinition>>,
    pub fact_vals: SecondaryMap<FactValId, Cow<'ast, ast::FieldDefinition>>,
    pub finish_funcs: SecondaryMap<FinishFuncId, Cow<'ast, AstNode<ast::FinishFunctionDefinition>>>,
    pub finish_func_args: SecondaryMap<FinishFuncArgId, Cow<'ast, ast::FieldDefinition>>,
    pub funcs: SecondaryMap<FuncId, Cow<'ast, AstNode<ast::FunctionDefinition>>>,
    pub func_args: SecondaryMap<FuncArgId, Cow<'ast, ast::FieldDefinition>>,
    pub global_lets: SecondaryMap<GlobalId, Cow<'ast, AstNode<ast::GlobalLetStatement>>>,
    pub structs: SecondaryMap<StructId, Cow<'ast, AstNode<ast::StructDefinition>>>,
    pub struct_fields:
        SecondaryMap<StructFieldId, Cow<'ast, ast::StructItem<ast::FieldDefinition>>>,
    pub stmts: SecondaryMap<StmtId, Cow<'ast, AstNode<ast::Statement>>>,
    pub exprs: SecondaryMap<ExprId, Cow<'ast, ast::Expression>>,
    pub idents: SecondaryMap<IdentId, Cow<'ast, ast::Identifier>>,
    pub blocks: SecondaryMap<BlockId, Cow<'ast, Vec<AstNode<ast::Statement>>>>,
    pub types: SecondaryMap<VTypeId, Cow<'ast, ast::VType>>,
}
