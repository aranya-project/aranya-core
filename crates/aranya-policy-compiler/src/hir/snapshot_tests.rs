#![cfg(test)]

use aranya_policy_ast::Version;
use aranya_policy_lang::lang::parse_policy_str;
use serde::Serialize;

use crate::hir::{self, arena::AstNodes, hir::Hir};

/// Wrapper struct for clean JSON snapshots
#[derive(Serialize)]
struct ParseSnapshot<'a> {
    hir: &'a Hir,
    ast_nodes: &'a AstNodes<'a>,
}

#[test]
fn test_parse_basic_policy() {
    let policy_text = r#"
action action1() {}
action action2(x int) {
    action action1()
    publish Cmd1 {}
}
action action3() {}

command Cmd1 {
    fields {}
    seal { return None }
    open { return None }
}

effect Effect1 {
    field1 int,
    field2 string,
}
effect Effect2 {}

// TODO: Add enum

fact Fact1[a int, b string]=>{c bool}
fact Fact2[]=>{}

// TODO: add function inputs
finish function func4() {}
finish function func5() {
    func4()
    delete Fact1[a: 42, b: "test"]
}

// TODO: add function inputs
function func1() int { return 42 }
function func2() bool { return true }
function func3() int { return func1() }

struct Struct1 {
    field1 bool,
    field2 int,
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    // Create snapshot of the parsed HIR and AST nodes
    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("basic_policy", &snapshot);
}

#[test]
fn test_parse_empty_policy() {
    let policy_text = "";
    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("empty_policy", &snapshot);
}

#[test]
fn test_parse_complex_expressions() {
    let policy_text = r#"
function complex_expr(x int, y int) bool {
    let sum = x + y
    let diff = x - y
    let is_positive = sum > 0
    let is_equal = x == y
    let result = is_positive && !is_equal
    return result || (diff < 0)
}

function optional_handling(x optional int) int {
    return unwrap x
}

function match_expr(x int) string {
    match x {
        1 => { return "one" }
        2 => { return "two" }
        _ => { return "other" }
    }
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("complex_expressions", &snapshot);
}

#[test]
fn test_parse_nested_structures() {
    let policy_text = r#"
struct Address {
    street string,
    city string,
}

struct Person {
    name string,
    age int,
    address struct Address,
}

command CreatePerson {
    fields {
        person struct Person,
    }
    seal { return None }
    open { return None }
}

action create_person(name string, age int, street string, city string) {
    let addr = Address {
        street: street,
        city: city,
    }
    let person = Person {
        name: name,
        age: age,
        address: addr,
    }
    publish CreatePerson { person: person }
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("nested_structures", &snapshot);
}

#[test]
fn test_parse_fact_operations() {
    let policy_text = r#"
fact User[user_id string]=>{name string, role string}
fact Permission[user_id string, resource string]=>{allowed bool}

action check_permissions(user_id string) {
    // map is valid in actions
    map Permission[user_id: user_id, resource: ?] as perm {
        check perm.allowed
    }
}

action check_admin(user_id string) {
    let user = query User[user_id: user_id]
    check user.role == "admin"
}

effect PermissionGranted {
    user_id string,
    resource string,
}

finish function grant_permission(user_id string, resource string) {
    create Permission[user_id: user_id, resource: resource]=>{allowed: true}
    // emit is valid in finish functions
    emit PermissionGranted {
        user_id: user_id,
        resource: resource,
    }
}

finish function revoke_permission(user_id string, resource string) {
    delete Permission[user_id: user_id, resource: resource]
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("fact_operations", &snapshot);
}
