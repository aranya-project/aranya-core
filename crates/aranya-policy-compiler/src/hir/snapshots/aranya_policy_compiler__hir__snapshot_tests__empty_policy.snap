---
source: crates/aranya-policy-compiler/src/hir/snapshot_tests.rs
expression: "&snapshot"
---
{
  "hir": {
    "actions": [
      {
        "value": null,
        "version": 0
      }
    ],
    "action_args": [
      {
        "value": null,
        "version": 0
      }
    ],
    "cmds": [
      {
        "value": null,
        "version": 0
      }
    ],
    "cmd_fields": [
      {
        "value": null,
        "version": 0
      }
    ],
    "effects": [
      {
        "value": null,
        "version": 0
      }
    ],
    "effect_fields": [
      {
        "value": null,
        "version": 0
      }
    ],
    "enums": [
      {
        "value": null,
        "version": 0
      }
    ],
    "facts": [
      {
        "value": null,
        "version": 0
      }
    ],
    "fact_keys": [
      {
        "value": null,
        "version": 0
      }
    ],
    "fact_vals": [
      {
        "value": null,
        "version": 0
      }
    ],
    "finish_funcs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "finish_func_args": [
      {
        "value": null,
        "version": 0
      }
    ],
    "funcs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "func_args": [
      {
        "value": null,
        "version": 0
      }
    ],
    "global_lets": [
      {
        "value": null,
        "version": 0
      }
    ],
    "structs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "struct_fields": [
      {
        "value": null,
        "version": 0
      }
    ],
    "stmts": [
      {
        "value": null,
        "version": 0
      }
    ],
    "exprs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "idents": [
      {
        "value": null,
        "version": 0
      }
    ],
    "blocks": [
      {
        "value": null,
        "version": 0
      }
    ],
    "types": [
      {
        "value": null,
        "version": 0
      }
    ],
    "ffi_imports": [
      {
        "value": null,
        "version": 0
      }
    ],
    "ffi_modules": [
      {
        "value": null,
        "version": 0
      }
    ],
    "ffi_funcs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "ffi_structs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "ffi_enums": [
      {
        "value": null,
        "version": 0
      }
    ]
  },
  "ast_nodes": {
    "actions": [
      {
        "value": null,
        "version": 0
      }
    ],
    "action_args": [
      {
        "value": null,
        "version": 0
      }
    ],
    "cmds": [
      {
        "value": null,
        "version": 0
      }
    ],
    "cmd_fields": [
      {
        "value": null,
        "version": 0
      }
    ],
    "effects": [
      {
        "value": null,
        "version": 0
      }
    ],
    "effect_fields": [
      {
        "value": null,
        "version": 0
      }
    ],
    "enums": [
      {
        "value": null,
        "version": 0
      }
    ],
    "facts": [
      {
        "value": null,
        "version": 0
      }
    ],
    "fact_keys": [
      {
        "value": null,
        "version": 0
      }
    ],
    "fact_vals": [
      {
        "value": null,
        "version": 0
      }
    ],
    "finish_funcs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "finish_func_args": [
      {
        "value": null,
        "version": 0
      }
    ],
    "funcs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "func_args": [
      {
        "value": null,
        "version": 0
      }
    ],
    "global_lets": [
      {
        "value": null,
        "version": 0
      }
    ],
    "structs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "struct_fields": [
      {
        "value": null,
        "version": 0
      }
    ],
    "stmts": [
      {
        "value": null,
        "version": 0
      }
    ],
    "exprs": [
      {
        "value": null,
        "version": 0
      }
    ],
    "idents": [
      {
        "value": null,
        "version": 0
      }
    ],
    "blocks": [
      {
        "value": null,
        "version": 0
      }
    ],
    "types": [
      {
        "value": null,
        "version": 0
      }
    ]
  }
}
